/*
* Set up: 
* $ go install google.golang.org/protobuf/cmd/protoc-gen-go@v1.26
* $ go install google.golang.org/grpc/cmd/protoc-gen-go-grpc@v1.1
*
* Add `<GOPATH>/bin` to your $PATH:
* OR (!suboptimal since it overwrites $PATH by appending an additional 
* line as opposed to change it directly)
* $ echo "export PATH=$PATH:$(go env GOPATH)/bin" >> ~/.profile 
* $ source ~/.profile 
* OR temporarily
* $ export PATH="$PATH:$(go env GOPATH)/bin"
*/
syntax = "proto3";

package sched;

option go_package = "schedproto/schedproto";

service Executor {
  // Remote execution via RPC call.
  rpc Execute (SchedRequest) returns (SchedReply) {}
  rpc ExecuteStream (stream SchedRequest) returns (SchedReply) {}
}

message SchedRequest {
  string schedAlg = 1; 
  string invocationName = 2;           // Text message field.
  uint32 batchsize = 3;         // batch size 
  uint32 runtimeInMilliSec = 4; // Execution runtime [ms].
  uint32 iterations = 5;        // remaining iterations 
  int32 deadline = 6;         // deadline 
  uint32 prevReplica = 7; // prevReplica
  int32 availableGPU = 8; // avaliable number of GPUs 
}

message SchedReply {
  repeated string invocationName = 1; 
  repeated uint32 replica = 2;   // The number of allocated GPUs 
  uint32 schedOverhead = 3;     // scheduling overhead 
}
